Add arch support for LoongArch64

enkerewpo opened the pull request:
1. add `loongarch64` assembly codes for unwinding support (https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html)
2. update the Github CI to include tests for the `loongarch64` target

enkerewpo committed:
Add basic framework for LoongArch64

enkerewpo committed:
Add basic loongarch64 context save and restore

enkerewpo committed:
Update CI configuration to include LoongArch64 target.

enkerewpo committed:
Allow unused unsafe warnings in LoongArch64 context save function

enkerewpo committed:
Locally tested the loongarch64 implementation using test_crates

enkerewpo committed:
Refactor imports and enhance LoongArch64 context save macro with additional register storage

enkerewpo committed:
Refactor LoongArch64 context save and restore macros to improve register handling and streamline assembly code generation

enkerewpo commented:
related PR discussions for using `unwinding` in `no_std` loongarch os kernels : https://github.com/asterinas/asterinas/pull/1915

enkerewpo commented:
Hi there,

I have tested the `loongarch64` port with both `std` and `no_std` test cases in `test_crates`, and updated the CI configuration to automatically run these tests on `loongarch64`.

For the bare-metal application tests on `loongarch64`, I created a minimal bare-metal example app ([repository link](https://github.com/enkerewpo/baremetal-loongarch64-unwinding-test)) to validate the `unwinding` library. The app was built with the following features:

```toml
[dependencies.unwinding]
path = "../unwinding" # my porting fork for loongarch64
default-features = false
features = ["fde-gnu-eh-frame-hdr", "panic", "personality", "unwinder"]
```

The bare-metal app successfully triggers a panic using `unwinding`’s logic, some important code sections:

**src/test.rs**  
```rust
// ... ...
pub fn test_panic() {
    println!("testing loongarch64 baremetal unwinding...");
    panic!("test panic for baremetal application using unwinding library");
}
// ... ...
```

**src/panic.rs**  
```rust
use core::ffi::c_void;
use gimli::Register;
use unwinding::abi::{
    _Unwind_Backtrace, _Unwind_FindEnclosingFunction, _Unwind_GetGR, _Unwind_GetIP, UnwindContext,
    UnwindReasonCode,
};

#[panic_handler]
fn panic(info: &core::panic::PanicInfo) -> ! {
    unsafe extern "Rust" {
        pub fn __panic_handler(info: &core::panic::PanicInfo) -> !;
    }
    unsafe { __panic_handler(info) }
}

#[unsafe(no_mangle)]
pub fn __panic_handler(info: &core::panic::PanicInfo) -> ! {
    println!(
        "(__panic_handler) panic at {:?}, reason: {:?}",
        info.location().unwrap(),
        info.message().as_str().unwrap_or("")
    );
    print_stack_trace();
    println!("halting, now we reach the end of the program");
    loop {}
}

pub fn print_stack_trace() {
    println!("printing stack trace...");

    struct CallbackData {
        counter: usize,
    }
    extern "C" fn callback(unwind_ctx: &UnwindContext<'_>, arg: *mut c_void) -> UnwindReasonCode {
        println!("callback...");
        let data = unsafe { &mut *(arg as *mut CallbackData) };
        data.counter += 1;
        let pc = _Unwind_GetIP(unwind_ctx);
        if pc > 0 {
            let fde_initial_address = _Unwind_FindEnclosingFunction(pc as *mut c_void) as usize;
            println!(
                "{:4}: fn {:#18x} - pc {:#18x} / registers:",
                data.counter, fde_initial_address, pc,
            );
        }
        // Print the first 8 general registers for any architecture. The register number follows
        // the DWARF standard.
        for i in 0..8u16 {
            let reg_i = _Unwind_GetGR(unwind_ctx, i as i32);
            // we are only testing on loongarch64 for now :D
            let reg_name = gimli::LoongArch::register_name(Register(i)).unwrap_or("unknown");
            if i % 4 == 0 {
                print!("\n    ");
            }
            print!(" {} {:#18x};", reg_name, reg_i);
        }
        print!("\n\n");
        UnwindReasonCode::NO_REASON
    }

    let mut data = CallbackData { counter: 0 };
    _Unwind_Backtrace(callback, &mut data as *mut _ as _);
}
```

**Test results:**  
The test application was executed on a `loongarch64` soft-float QEMU environment. The panic handling and stack unwinding completed successfully, producing the expected output.

<img width="1435" height="624" alt="image" src="https://github.com/user-attachments/assets/6da9a8e4-70ac-4444-888f-233d0b445231" />

To reproduce this result:

```bash
# install qemu-system-loongarch64 through package manager or just compile it from official QEMU git repo
# note: ./configure --target-list="loongarch64-softmmu"
git clone https://github.com/enkerewpo/unwinding
git clone https://github.com/enkerewpo/baremetal-loongarch64-unwinding-test
cd baremetal-loongarch64-unwinding-test
./build.sh
./run_qemu.sh
```

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,247 @@
+use super::maybe_cfi;
+
+// https://github.com/loongson/la-abi-specs/blob/release/ladwarf.adoc
+pub const MAX_REG_RULES: usize = 65;
```
Why is this 65? Looks it should be 64 from the doc?

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,247 @@
+                &format_args!("{:#x}", self.gp[i]),
+            );
+        }
+        fmt.field("sp", &self.gp[3]);
```
This isn't needed?

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,247 @@
+            "
+            st.d $r1, $r3, 0x200
+            ",
+            maybe_cfi!(".cfi_offset 1, -16"), // ra
```
Does loongarch assembler not support more friendly names in `cfi_offset`?

nbdd0121 replied:
Have you tried `.cfi_offset $ra, -16` (note the $ sign)

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,247 @@
+    };
+}
+
+macro_rules! helper {
```
I think this doesn't have to be a separate macro_rules.

enkerewpo committed:
Remove helper macro and `sp` in format output

enkerewpo commented:
For the current number of `DWARF_FRAME_REGISTERS`, I just checked the latest GCC code(https://github.com/gcc-mirror/gcc) and its loongarch config. Since `DWARF_FRAME_REGISTERS` is not defined in `gcc/config/loongarch`, GCC will set it to `FIRST_PSEUDO_REGISTER` (according to `gcc/defaults.h`) which is defined as 74 in `gcc/config/loongarch/loongarch.h`. I'll update the `MAX_REG_RULES` to this number.

https://github.com/gcc-mirror/gcc/blob/48ef4af2274a0a1ad945ac14217806c1c1d8ccb2/gcc/config/loongarch/loongarch.h#L306

https://github.com/gcc-mirror/gcc/blob/48ef4af2274a0a1ad945ac14217806c1c1d8ccb2/gcc/defaults.h#L409

enkerewpo commented:
For the `.cfi_offset`, LLVM still does not support fancy names here **for loongarch**, a quick test on LLVM 19:
<img width="1372" height="945" alt="image" src="https://github.com/user-attachments/assets/57f6994e-d75e-452a-8e45-6e6be2f8de76" />

enkerewpo committed:
Update MAX_REG_RULES constant for LoongArch64

enkerewpo committed:
Merge branch 'nbdd0121:trunk' into trunk

enkerewpo commented:
I just found a problem, when using `begin_panic` instead of `panic!`, the panic process will trigger endless panics of "pointer mis-alignment":

<img width="1403" height="938" alt="image" src="https://github.com/user-attachments/assets/09855e7e-4c68-4dd3-ae93-cdf714b2de9e" />

```
...
(__panic_handler) panic at Location { file: "/home/wheatfox/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/manually_drop.rs", line: 223, column: 18 }, reason: "unsafe precondition(s) violated: ptr::read requires that the pointer argument is aligned and non-null\n\nThis indicates a bug in the program. This Undefined Behavior check is optional, and cannot be relied on for safety."
printing stack trace...
(__panic_handler) panic at Location { file: "/home/wheatfox/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/manually_drop.rs", line: 223, column: 18 }, reason: "unsafe precondition(s) violated: ptr::read requires that the pointer argument is aligned and non-null\n\nThis indicates a bug in the program. This Undefined Behavior check is optional, and cannot be relied on for safety."
printing stack trace...
(__panic_handler) panic at Location { file: "/home/wheatfox/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/manually_drop.rs", line: 223, column: 18 }, reason: "unsafe precondition(s) violated: ptr::read requires that the pointer argument is aligned and non-null\n\nThis indicates a bug in the program. This Undefined Behavior check is optional, and cannot be relied on for safety."
printing stack trace...
...
```

I'll look into this problem.

enkerewpo commented:
Hi there,

I fixed the problem, in the linker script of my loongarch64 baremetal program, I wrongly set the inital `sp` to a non-8B aligned address : ( . After fixing the linker script, **everything works fine** using `unwinding`'s `catch_unwind` and `begin_panic` in a loongarch64 baremetal environment (`qemu-system-loongarch64`):

```rust
// ...
    uart_init();
    heap_init();
    println!("hello world! {}", 42);

    catch_unwind(|| {
        println!("testing catch unwind...");
        test_panic();
        println!("testing catch unwind done");
    })
    .unwrap();

    // set global trap handler
    set_global_trap_handler(trap_handler);

    // test_vtimer();

    // panic!("test panic! (rust)"); // uncomment this when testing `panic!`

    println!("going to idle loop, the main function is done");
// ...
```

```rust
use unwinding::panic::begin_panic;

pub fn test_panic() {
    // panic!("test panic");
    begin_panic(Box::new("test panic"));
}
```
1. Testing the unwinding process:

<img width="1122" height="791" alt="image" src="https://github.com/user-attachments/assets/d24bb3d0-89ed-4108-9371-d1f121eeb647" />

2. Testing the `panic!` and stacktrace dump:
<img width="1122" height="724" alt="image" src="https://github.com/user-attachments/assets/5de61137-1729-434c-aa16-92ccfd768872" />

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,228 @@
+#[derive(Clone, Default)]
+pub struct Context {
+    pub gp: [usize; 32],
+    pub fp: [usize; 32],
```
This could be optional depending on target_feature? See riscv64.

Is the idea to avoid having different size of `Context` to simplify assembly?

enkerewpo replied:
I'll add the target_feature like riscv

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,228 @@
+        core::arch::naked_asm!(
+            maybe_cfi!(".cfi_startproc"),
+            "
+            move $r12, $r3
```
Does `move $t0, $sp` not work?

enkerewpo replied:
I just realized that Rust finally add support for using loongarch reg names like `ra` and `sp` in `asm!` stuff (not supported in the older versions of Rust compilers) 
<img width="855" height="140" alt="image" src="https://github.com/user-attachments/assets/9892d1fb-c2bd-4097-83e0-4ecdeea2d6ee" />

I will update these to reg names.

nbdd0121 replied:
These names are not handled by Rust, they're handled by LLVM directly

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,228 @@
+            st.d $r0, $r3, 0x0 // zero
+            st.d $r1, $r3, 0x8 // ra
+            st.d $r2, $r3, 0x10 // tp
+            // sp is saved later
```
Any reason to defer the saving here? The value is in `$t0` now, so you can avoid recalculating it later. See how rv64 code does this.

enkerewpo replied:
ok

enkerewpo commented:
Thanks for your comments! I updated the related code and tested the hard-float functionality in loongarch64 qemu with rust target `loongarch64-unknown-none` (the previous tests are using `loongarch64-unknown-none-softfloat` which does not have the `d` feature).

Below is a catch+unwind with `panic!`+stacktrace (with reg dumps for both GPRs, and FPs, which begin with `f`):

<img width="846" height="349" alt="image" src="https://github.com/user-attachments/assets/adbf7ebf-b78f-4166-866c-18f2a62f402f" />

However, I found that the `MAX_REG_RULES` must be set to values <= 64. Setting it to above 64 will cause the unwinding to fail (for some reason):

<img width="852" height="285" alt="image" src="https://github.com/user-attachments/assets/f32babb9-75af-43bb-8c4f-0c4dd2515d3c" />

UPDATE: I fixed the problem mentioned above (which actually stuck at creating the unwinding context). It turns out that the **stack size** I set for my baremetal program is *not big enough*. After doubling the stack size, eveything works fine.

enkerewpo committed:
Enhance context handling with conditional FP support in assembly macros

nbdd0121 commented:
Starting in LLVM 20, CFI directive can use register names: https://github.com/llvm/llvm-project/pull/117120

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,229 @@
+        for i in 0..=31 {
+            fmt.field(
+                LoongArch::register_name(Register(i as _)).unwrap(),
+                &format_args!("{:#x}", self.gp[i]),
```
Please leave out `format_args` to make this arch consistent with others. If hex is needed, one can use `x?` to do a hex debug.

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,229 @@
+use core::{fmt, ops};
```
This project uses module-level granularity. Please split these.

nbdd0121 commented on src/unwinder/arch/loongarch64.rs:
```diff
@@ -0,0 +1,229 @@
+            maybe_cfi!(".cfi_startproc"),
+            "
+            move $t0, $sp
+            addi.d $sp, $sp, -0x210
```
This uses more stack space than necessary if fp is not present. I'd still recommend RISC-V-like way of organise the macros so that the offsets can be different for different target feature combinations.

nbdd0121 reviewed (commented):
Sorry for the late review.. This was completely lost from my radar.

enkerewpo committed:
Merge branch 'nbdd0121:trunk' into trunk

enkerewpo committed:
Refactor LoongArch64 context save and restore macros for improved clarity and stack space efficiency.

enkerewpo commented:
OK :D

nbdd0121 commented:
Thank you!
